File: java/src/test/java/org/cassandragargoyle/api/fuzzy/LevenshteinTest.java
/*
 *  This file is part of CassandraGargoyle Community Project
 *  Licensed under the MIT License - see LICENSE file for details
 */
package org.cassandragargoyle.api.fuzzy;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import org.junit.jupiter.api.Test;

/**
 * Tests for {@link Levenshtein}.
 *
 * <p>
 * The {@code shouldMatchSharedVector*} tests cover the canonical vectors
 * defined in issue #010 and must agree byte-for-byte with the Python and Go
 * implementations of the same package.
 *
 * @author Zdenek
 */
public class LevenshteinTest
{
	private static final double EPS = 1e-9;

	@Test
	public void shouldMatchSharedVectorBothEmpty()
	{
		assertEquals(0, Levenshtein.distance("", ""));
		assertEquals(1.0, Levenshtein.similarity("", ""), EPS);
	}

	@Test
	public void shouldMatchSharedVectorFirstEmpty()
	{
		assertEquals(3, Levenshtein.distance("", "abc"));
		assertEquals(0.0, Levenshtein.similarity("", "abc"), EPS);
	}

	@Test
	public void shouldMatchSharedVectorSecondEmpty()
	{
		assertEquals(3, Levenshtein.distance("abc", ""));
		assertEquals(0.0, Levenshtein.similarity("abc", ""), EPS);
	}

	@Test
	public void shouldMatchSharedVectorIdentical()
	{
		assertEquals(0, Levenshtein.distance("abc", "abc"));
		assertEquals(1.0, Levenshtein.similarity("abc", "abc"), EPS);
	}

	@Test
	public void shouldMatchSharedVectorKittenSitting()
	{
		assertEquals(3, Levenshtein.distance("kitten", "sitting"));
		assertEquals(1.0 - 3.0 / 7.0, Levenshtein.similarity("kitten", "sitting"), EPS);
	}

	@Test
	public void shouldMatchSharedVectorFlawLawn()
	{
		assertEquals(2, Levenshtein.distance("flaw", "lawn"));
		assertEquals(0.5, Levenshtein.similarity("flaw", "lawn"), EPS);
	}

	@Test
	public void shouldMatchSharedVectorGumboGambol()
	{
		assertEquals(2, Levenshtein.distance("gumbo", "gambol"));
		assertEquals(1.0 - 2.0 / 6.0, Levenshtein.similarity("gumbo", "gambol"), EPS);
	}

	@Test
	public void shouldMatchSharedVectorSaturdaySunday()
	{
		assertEquals(3, Levenshtein.distance("Saturday", "Sunday"));
		assertEquals(1.0 - 3.0 / 8.0, Levenshtein.similarity("Saturday", "Sunday"), EPS);
	}

	@Test
	public void shouldMatchSharedVectorCafeAccented()
	{
		assertEquals(1, Levenshtein.distance("café", "cafe"));
		assertEquals(0.75, Levenshtein.similarity("café", "cafe"), EPS);
	}

	@Test
	public void shouldMatchSharedVectorCjkPrefix()
	{
		assertEquals(1, Levenshtein.distance("日本語", "日本"));
		assertEquals(1.0 - 1.0 / 3.0, Levenshtein.similarity("日本語", "日本"), EPS);
	}

	@Test
	public void shouldBeSymmetric()
	{
		assertEquals(
			Levenshtein.distance("kitten", "sitting"),
			Levenshtein.distance("sitting", "kitten"));
		assertEquals(
			Levenshtein.distance("café", "cafe"),
			Levenshtein.distance("cafe", "café"));
	}

	@Test
	public void shouldCountCodePointsNotUtf16Units()
	{
		// U+1F600 GRINNING FACE is a supplementary code point encoded as a UTF-16
		// surrogate pair, i.e. String#length() reports 2. distance("😀", "")
		// must therefore be 1, not 2.
		String emoji = "😀";
		assertEquals(2, emoji.length());
		assertEquals(1, Levenshtein.distance(emoji, ""));
		assertEquals(0.0, Levenshtein.similarity(emoji, ""), EPS);
		assertEquals(1, Levenshtein.distance("", emoji));
	}

	@Test
	public void shouldRejectNullInputs()
	{
		assertThrows(NullPointerException.class, () -> Levenshtein.distance(null, "x"));
		assertThrows(NullPointerException.class, () -> Levenshtein.distance("x", null));
		assertThrows(NullPointerException.class, () -> Levenshtein.similarity(null, "x"));
		assertThrows(NullPointerException.class, () -> Levenshtein.similarity("x", null));
	}

	@Test
	public void similarityIsBounded()
	{
		double s = Levenshtein.similarity("kitten", "sitting");
		assertTrue(s >= 0.0 && s <= 1.0);
	}
}

File: go/fuzzy/levenshtein_test.go
// This file is part of CassandraGargoyle Community Project
// Licensed under the MIT License - see LICENSE file for details

package fuzzy

import (
	"math"
	"testing"
)

const eps = 1e-9

// TestDistanceSharedVectors covers the canonical vectors from issue #010.
// These cases must agree byte-for-byte with the Java and Python implementations.
func TestDistanceSharedVectors(t *testing.T) {
	tests := []struct {
		name string
		a, b string
		want int
	}{
		{"both empty", "", "", 0},
		{"first empty", "", "abc", 3},
		{"second empty", "abc", "", 3},
		{"identical", "abc", "abc", 0},
		{"kitten vs sitting", "kitten", "sitting", 3},
		{"flaw vs lawn", "flaw", "lawn", 2},
		{"gumbo vs gambol", "gumbo", "gambol", 2},
		{"Saturday vs Sunday", "Saturday", "Sunday", 3},
		{"café vs cafe", "café", "cafe", 1},
		{"cjk prefix", "日本語", "日本", 1},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := Distance(tt.a, tt.b)
			if got != tt.want {
				t.Errorf("Distance(%q, %q) = %d, want %d", tt.a, tt.b, got, tt.want)
			}
		})
	}
}

func TestSimilaritySharedVectors(t *testing.T) {
	tests := []struct {
		name string
		a, b string
		want float64
	}{
		{"both empty", "", "", 1.0},
		{"first empty", "", "abc", 0.0},
		{"second empty", "abc", "", 0.0},
		{"identical", "abc", "abc", 1.0},
		{"kitten vs sitting", "kitten", "sitting", 1.0 - 3.0/7.0},
		{"flaw vs lawn", "flaw", "lawn", 0.5},
		{"gumbo vs gambol", "gumbo", "gambol", 1.0 - 2.0/6.0},
		{"Saturday vs Sunday", "Saturday", "Sunday", 1.0 - 3.0/8.0},
		{"café vs cafe", "café", "cafe", 0.75},
		{"cjk prefix", "日本語", "日本", 1.0 - 1.0/3.0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := Similarity(tt.a, tt.b)
			if math.Abs(got-tt.want) > eps {
				t.Errorf("Similarity(%q, %q) = %v, want %v", tt.a, tt.b, got, tt.want)
			}
		})
	}
}

func TestDistanceIsSymmetric(t *testing.T) {
	pairs := []struct{ a, b string }{
		{"kitten", "sitting"},
		{"café", "cafe"},
		{"日本語", "日本"},
	}
	for _, p := range pairs {
		t.Run(p.a+"<->"+p.b, func(t *testing.T) {
			ab := Distance(p.a, p.b)
			ba := Distance(p.b, p.a)
			if ab != ba {
				t.Errorf("symmetry broken: Distance(%q, %q)=%d but Distance(%q, %q)=%d",
					p.a, p.b, ab, p.b, p.a, ba)
			}
		})
	}
}

// TestRuneVsByteLength is the headline regression case for the Go binding.
// "café" is 4 runes but 5 UTF-8 bytes (the é is 0xC3 0xA9). A byte-based
// implementation would report distance("café", "cafe") = 2 — wrong.
func TestRuneVsByteLength(t *testing.T) {
	if got := Distance("café", "cafe"); got != 1 {
		t.Errorf("Distance(\"café\", \"cafe\") = %d, want 1 — implementation operates on bytes, not runes", got)
	}
	// Emoji is a 4-byte UTF-8 sequence representing a single rune.
	if got := Distance("😀", ""); got != 1 {
		t.Errorf("Distance(\"😀\", \"\") = %d, want 1", got)
	}
}

func TestSimilarityIsBounded(t *testing.T) {
	s := Similarity("kitten", "sitting")
	if s < 0.0 || s > 1.0 {
		t.Errorf("Similarity out of range: %v", s)
	}
}
